core/shared/src/main/scala/scalaz/zio/Retry.scala

```diff
@@ -0,0 +1,126 @@
+   */
+  def update(e: E, s: S): IO[E, S]
+
+  /**
+   * Peeks at the state, executes some action, and then continues retrying or not
+   * based on the specified predicate.
+   */
+  final def check[A](action: (E, S) => IO[E, A])(pred: A => Boolean): Retry[S, E] =
+    new Retry[S, E] {
+      val initial = self.initial
+
+      def update(e: E, s: S): IO[E, S] =
+        for {
+          s <- self.update(e, s)
+          a <- action(e, s)
+          _ <- if (pred(a)) IO.now(s) else IO.fail(e)
+        } yield s
+    }
+
+  /**
+   * Returns a new strategy that retries while the error matches the condition.
+   */
+  final def whileError(p: E => Boolean): Retry[S, E] =
+    check[E]((e, _) => IO.now(e))(p)
+
+  /**
+   * Returns a new strategy that retries until the error matches the condition.
+   */
+  final def untilError(p: E => Boolean): Retry[S, E] =
+    whileError(negate(p))
+
+  /*
+   * Returns a new strategy that retries until the state matches the condition.
+   */
+  final def untilState(p: S => Boolean): Retry[S, E] =
+    check[S]((_, s) => IO.now(s))(p)
+
+  /*
+   * Returns a new strategy that retries while the state matches the condition.
+   */
+  final def whileState(p: S => Boolean): Retry[S, E] =
+    untilState(negate(p))
+
+  /**
+   * Returns a new strategy that retries for as long as this strategy and the
+   * specified strategy both agree to retry.
+   */
+  final def intersect[S2](that: Retry[S2, E]): Retry[(S, S2), E] =
+    new Retry[(S, S2), E] {
+      val initial = self.initial.zip(that.initial)
+
+      def update(e: E, s: (S, S2)): IO[E, (S, S2)] =
+        self.update(e, s._1).zip(that.update(e, s._2))
+    }
+
+  final def unionLeft(that: Retry[S, E]): Retry[S, E] =
+    self.unionWith(that)((s, _) => s)
+
+  final def unionRight(that: Retry[S, E]): Retry[S, E] =
+    self.unionWith(that)((_, s) => s)
+
+  /**
+   * Returns a new strategy that retries for as long as either this strategy or
+   * the specified strategy agree to retry.
+   */
+  final def unionWith(that: Retry[S, E])(f: (S, S) => S): Retry[S, E] =
+    new Retry[S, E] {
+      val initial = self.initial.zipWith(that.initial)(f)
+
+      def update(e: E, s: S): IO[E, S] =
+        (self update (e, s)) orElse (that update (e, s))
+    }
+
+  private def negate[A](f: A => Boolean): A => Boolean =
+    (a: A) => !f(a)
+}
+
+object Retry {
+  final def apply[S, E](initial0: IO[E, S], update0: (E, S) => IO[E, S]): Retry[S, E] = new Retry[S, E] {
+    val initial                      = initial0
+    def update(e: E, s: S): IO[E, S] = update0(e, s)
+  }
+
+  /**
+   * A retry strategy that always retries and counts the number of retries.
+   */
+  final def counted[E]: Retry[Int, E] =
+    Retry[Int, E](IO.now(0), (_, i) => IO.now(i + 1))
+
+  /**
+   * A retry strategy that always retries and computes the time since the
+   * beginning of the process.
+   */
+  final def timed[E]: Retry[(Long, Long), E] = {
```

Comment:
This is the one thing I don't like. All the state has to be exposed via `S`, which means you get this nasty tuple here that carries a piece of state (the starting time) that I don't want to expose. Maybe it'd be interesting to look at having a part of the state hidden, and then a projection from the hidden state `S0 => S` (to the visible part, which would be exposed in functions like `whileState`).